internal/exec/util/file.go

```diff
@@ -152,19 +152,23 @@ func (u Util) WriteLink(s types.Link) error {
 
 func (u Util) SetPermissions(mode *int, node types.Node) error {
 	if mode != nil {
+		// Set ownership and then permissions, https://man7.org/linux/man-pages/man2/lchown.2.html
+		// "... When the owner or group of an executable file is changed by an
+		// unprivileged user, the S_ISUID and S_ISGID mode bits are cleared..."
+
+		defaultUid, defaultGid, _ := getFileOwnerAndMode(node.Path)
+		uid, gid, err := u.ResolveNodeUidAndGid(node, defaultUid, defaultGid)
+		if err != nil {
+			return fmt.Errorf("failed to determine correct uid and gid for %s: %v", node.Path, err)
+		}
+		if err := os.Lchown(node.Path, uid, gid); err != nil {
+			return fmt.Errorf("failed to change ownership of %s: %v", node.Path, err)
+		}
```

Comment:
That should probably be moved above out of the mode check

Reply:
Ah yeah makes sense.

Reply:
Hum, I meant the code that does the chown. We still want to chown even if we don't change the mode.

Reply:
Oh, yes, that would make sense; lol sorry bad assumption.